File: src/agv_common_library/src/module_manager.cpp
#include <agv_common_library/module_manager.h>

namespace module_manager
{
ModuleServer::ModuleServer(ros::NodeHandle nh, const std::string& name)
    : nh_(nh), name_(name), module_status(ModuleStatus::WAITING)
{
  module_status_pub = nh_.advertise<std_stamped_msgs::StringStamped>(
      name_ + "/module_status", 5);

  run_pause_req_sub =
      nh_.subscribe(name_ + "/run_pause_req", 10,
                    &ModuleServer::runPauseRequestCallback, this);
  reset_error_sub = nh_.subscribe(name_ + "/reset_error", 10,
                                  &ModuleServer::resetErrorCallback, this);
  request_start_mission_sub =
      nh_.subscribe("/request_start_mission", 10,
                    &ModuleServer::requestStartMissionCallback, this);
  odom_sub = nh_.subscribe("/odom", 10, &ModuleServer::odomCallback, this);

  error_code = "";
  reset_action_req = false;
  reset_error_request = false;
  resume_req = false;
  pause_req = false;
  resume_req_by_server = false;
  pause_req_by_server = false;
  action_running = false;
  module_state = "";
}

ModuleServer::~ModuleServer()
{
  // Destructor implementation, if needed
}

void ModuleServer::startROS()
{
    ros::spinOnce();
}

void ModuleServer::odomCallback(const nav_msgs::Odometry::ConstPtr& msg)
{
  vel_x = msg->twist.twist.linear.x;
}

void ModuleServer::requestStartMissionCallback(
    const std_stamped_msgs::StringStamped::ConstPtr& msg)
{
  if (msg->data == "STOP")
  {
    reset_action_req = true;
    ROS_INFO("STOP MISSION");
  }
}

void ModuleServer::runPauseRequestCallback(
    const std_stamped_msgs::StringStamped::ConstPtr& msg)
{
  if (msg->data == "RUN")
  {
    ROS_INFO("Resume request");
    resume_req = true;
    pause_req = false;
    resume_req_by_server = true;
    pause_req_by_server = false;
  }
  else if (msg->data == "PAUSE")
  {
    ROS_INFO("Pause request");
    resume_req = false;
    pause_req = true;
  }
  else if (msg->data == "PAUSE_BY_SERVER")
  {
    ROS_INFO("Pause request");
    resume_req = false;
    pause_req = true;
    resume_req_by_server = false;
    pause_req_by_server = true;
  }
}

void ModuleServer::resetErrorCallback(
    const std_stamped_msgs::EmptyStamped::ConstPtr& msg)
{
  reset_error_request = true;
}

void ModuleServer::resetFlags()
{
  pause_req = false;
  resume_req = false;
  resume_req_by_server = false;
  pause_req_by_server = false;
  reset_action_req = false;
  reset_error_request = false;
  error_code = "";
}

void ModuleServer::sendFeedback(const ros::Publisher& action,
                                const std::string& msg)
{
  std_stamped_msgs::StringStamped feedback_msg;
  feedback_msg.data = msg;
  action.publish(feedback_msg);
}

ModuleClient::ModuleClient(ros::NodeHandle nh, const std::string& name,
                           const YAML::Node& param_dict)
    : nh_(nh), name_(name), param_dict_(param_dict), parse_json(true),
      handle_when_sim(true), timeout(2.0), pause_if_error(true),
      module_alive(false), check_script("")
{
  ROS_INFO("Init ModuleClient for: %s", name_.c_str());

  if (param_dict.IsMap())
  {
    ROS_INFO("Param_dict is Map");
    if (param_dict["display_name"])
    {
      display_name = param_dict["display_name"].as<std::string>();
    }
    if (param_dict["parse_json"])
    {
      parse_json = param_dict["parse_json"].as<bool>();
    }
    if (param_dict["handle_when_sim"])
    {
      handle_when_sim = param_dict["handle_when_sim"].as<bool>();
    }
    if (param_dict["timeout"])
    {
      timeout = param_dict["timeout"].as<double>();
    }
    if (param_dict["pause_if_error"])
    {
      pause_if_error = param_dict["pause_if_error"].as<bool>();
    }
    if (param_dict["check_script"])
    {
      check_script = param_dict["check_script"].as<std::string>();
    }

    if (param_dict["error_list"])
    {
      for (const auto& error_node : param_dict["error_list"])
      {
        for (const auto& it : error_node)
        {
          std::string error_type = it.first.as<std::string>();
          std::map<std::string, std::string> details_map;
          for (const auto& detail : it.second)
          {
            details_map[detail.first.as<std::string>()] =
                detail.second.as<std::string>();
          }
          error_list.push_back({{error_type, details_map}});
        }
      }
    }

    if (param_dict["state_list"])
    {
      for (const auto& error_node : param_dict["state_list"])
      {
        for (const auto& it : error_node)
        {
          std::string error_type = it.first.as<std::string>();
          std::map<std::string, std::string> details_map;
          for (const auto& detail : it.second)
          {
            details_map[detail.first.as<std::string>()] =
                detail.second.as<std::string>();
          }
          state_list.push_back({{error_type, details_map}});
        }
      }
    }
  }
  else
  {
    ROS_INFO("Param_dict not is Map");
  }
  last_module_status = ros::Time::now().toSec();
  module_alive = false;

  setupSubscriber();

  loop_thread = std::thread(&ModuleClient::loop, this);
}

ModuleClient::~ModuleClient()
{
  if (loop_thread.joinable())
  {
    loop_thread.join();
  }
}

void ModuleClient::setupSubscriber()
{
  std::string status_topic = (param_dict_["status_topic"])
                                 ? param_dict_["status_topic"].as<std::string>()
                                 : (name_ + "/module_status");
  module_status_sub = nh_.subscribe(status_topic, 10,
                                    &ModuleClient::moduleStatusCallback, this);
  ROS_INFO("Init Subcriber topic for: %s", name_.c_str());
}

void ModuleClient::moduleStatusCallback(
    const std_stamped_msgs::StringStamped::ConstPtr& msg)
{
  module_status_msg = msg;
  last_module_status = ros::Time::now().toSec();
  if (!parse_json)
  {
    return;
  }
  special_matching_error.clear();
  special_matching_state.clear();
  special_error_led.clear();
  special_state_led.clear();
  special_state_sound.clear();
  special_error_sound.clear();

  try
  {
    processStatusMessage(msg->data);
  }
  catch (const std::exception& e)
  {
    ROS_ERROR("module_status_cb: %s", e.what());
  }
}

void ModuleClient::processStatusMessage(const std::string& data)
{
  Json::Reader reader;
  Json::Value root;
  if (!reader.parse(data, root))
  {
    ROS_ERROR("Failed to parse JSON");
    return;
  }
  module_status_dict.clear();
  for (Json::Value::const_iterator it = root.begin(); it != root.end(); ++it)
  {
    module_status_dict[it.key().asString()] = it->asString();
  }

  if (module_status_dict.count("error_code"))
  {
    error_code = module_status_dict["error_code"];
  }

  // Handle error list
  if (!error_code.empty())
  {
    for (const auto& error : error_list)
    {
      if (error.count(error_code))
      {
        special_matching_error = error.at(error_code).at("detail_status");
        if (error.at(error_code).count("led_status"))
        {
          special_error_led = error.at(error_code).at("led_status");
        }
        if (error.at(error_code).count("sound_status"))
        {
          special_error_sound = error.at(error_code).at("sound_status");
        }
        return;
      }
    }
  }

  // Handle state list
  if (module_status_dict.count("state"))
  {
    std::string state = module_status_dict["state"];
    for (const auto& state_item : state_list)
    {
      if (state_item.count(state))
      {
        special_matching_state = state_item.at(state).at("detail_status");
        if (state_item.at(state).count("led_status"))
        {
          special_state_led = state_item.at(state).at("led_status");
        }
        if (state_item.at(state).count("sound_status"))
        {
          special_state_sound = state_item.at(state).at("sound_status");
        }
        return;
      }
    }
  }
}

void ModuleClient::loop()
{
  ros::Rate rate(2);
  double vel_x = 0;
  double start_time_check_disconnect = ros::Time::now().toSec();

  while (ros::ok())
  {
    if (param_dict_["status_topic"] &&
        param_dict_["status_topic"].as<std::string>() == "/followline_sensor")
    {
      if (vel_x == 0)
      {
        start_time_check_disconnect = ros::Time::now().toSec();
        last_module_status = ros::Time::now().toSec();
      }
      if (ros::Time::now().toSec() - start_time_check_disconnect <= 0.5)
      {
        last_module_status = ros::Time::now().toSec();
      }
    }
    if (check_script.empty())
    {
      module_alive = (ros::Time::now().toSec() - last_module_status <= timeout);
    }
    else
    {
      // Implement custom check script logic here
    }
    rate.sleep();
  }
}

}  // namespace module_manager

File: src/agv_common_library/scripts/module_manager.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys

python3 = True if sys.hexversion > 0x03000000 else False
import rospy
import json
import threading
from std_stamped_msgs.msg import StringStamped, EmptyStamped
from common_function import EnumString
from nav_msgs.msg import Odometry

class ModuleStatus(EnumString):
    WAITING = 0
    RUNNING = 1
    PAUSED = 2
    ERROR = 3


class ModuleServer:
    def __init__(self, name):
        # Ros
        self.module_status_pub = rospy.Publisher(
            name + "/module_status", StringStamped, queue_size=5
        )
        rospy.Subscriber(
            name + "/run_pause_req", StringStamped, self.run_pause_req_cb
        )
        rospy.Subscriber(
            name + "/reset_error", EmptyStamped, self.reset_error_cb
        )
        rospy.Subscriber(
            "/request_start_mission",
            StringStamped,
            self.request_start_mission_cb,
        )
        rospy.Subscriber("/odom", Odometry, self.odom_cb)
        # Variable
        self.name = name
        self.error_code = ""
        self.reset_action_req = False
        self.reset_error_request = False
        self.resume_req = False
        self.pause_req = False
        self.resume_req_by_server = False
        self.pause_req_by_server = False
        self.action_running = False
        self.module_state = ""
        self.module_status = ModuleStatus.WAITING
        self.param_dict = {}

    def odom_cb(self, msg):
        self.vel_x = msg.twist.twist.linear.x

    def request_start_mission_cb(self, msg):
        if msg.data == "STOP":
            self.reset_action_req = True
            rospy.loginfo("STOP MISSION")

    def run_pause_req_cb(self, msg):
        if msg.data == "RUN":
            rospy.loginfo("Resume request")
            self.resume_req = True
            self.pause_req = False
            self.resume_req_by_server = True
            self.pause_req_by_server = False
        elif msg.data == "PAUSE":
            rospy.loginfo("Pause request")
            self.resume_req = False
            self.pause_req = True
            self.resume_req_by_server = False
            self.pause_req_by_server = False
        elif msg.data == "PAUSE_BY_SERVER":
            rospy.loginfo("Pause request")
            self.resume_req = False
            self.pause_req = True
            self.resume_req_by_server = False
            self.pause_req_by_server = True

    def reset_error_cb(self, msg):
        self.reset_error_request = True

    def reset_flag(self):
        self.pause_req = False
        self.resume_req = False
        self.resume_req_by_server = False
        self.pause_req_by_server = False
        self.reset_action_req = False
        self.reset_error_request = False
        self.error_code = ""

    def send_feedback(self, action, msg):
        self._feedback.data = msg
        action.publish_feedback(self._feedback)


class ModuleClient:
    def __init__(self, name, param_dict={}):
        self.param_dict = param_dict
        rospy.loginfo("Init ModuleClient for: {}".format(name))
        rospy.Subscriber("/odom", Odometry, self.odom_cb)
        # Ros
        if not "status_topic" in param_dict:
            rospy.Subscriber(
                name + "/module_status", StringStamped, self.module_status_cb
            )
        else:
            [msg_ns, msg_type] = param_dict["status_msg"].split("/", 1)
            cmd = "from {}.msg import {}".format(msg_ns, msg_type)
            exec(cmd)
            cmd = 'rospy.Subscriber("{}", {}, self.module_status_cb)'.format(
                param_dict["status_topic"], msg_type
            )
            exec(cmd)

        # Variable
        self.name = name
        self.display_name = param_dict["display_name"]
        self.parse_json = (
            param_dict["parse_json"] if "parse_json" in param_dict else True
        )
        self.handle_when_sim = (
            param_dict["handle_when_sim"]
            if "handle_when_sim" in param_dict
            else True
        )
        self.condition = param_dict["condition"] if "condition" in param_dict else {}
        self.type_skip = self.condition.get("type", "")
        self.skip_threshold = self.condition.get("threshold", 0.0)


        self.timeout = param_dict["timeout"] if "timeout" in param_dict else 2.0
        self.pause_if_error = (
            param_dict["pause_if_error"]
            if "pause_if_error" in param_dict
            else True
        )
        self.error_code_number_default = (
            param_dict["error_code_number_default"]
            if "error_code_number_default" in param_dict
            else -1
        )
        self.reset_when_mission_error = True
        self.display_error = (
            param_dict["display_error"]
            if "display_error" in param_dict
            else True
        )
        self.error_list = (
            param_dict["error_list"] if "error_list" in param_dict else []
        )
        self.state_list = (
            param_dict["state_list"] if "state_list" in param_dict else []
        )
        self.check_script = (
            param_dict["check_script"] if "check_script" in param_dict else ""
        )

        self.last_module_status = rospy.get_time()
        self.module_alive = False
        self.module_status_msg = None
        self.current_vel_x = 0.0
        self.map_save_thread = threading.Thread(target=self.loop, args=())
        self.map_save_thread.start()
        self.module_status_dict = {}
        self.special_matching_error = ""
        self.special_matching_state = ""
        self.special_error_led = ""
        self.special_state_led = ""
        self.special_state_sound = ""
        self.special_error_sound = ""
        self.error_code = ""
        self.error_code_number = -1  # not define
    def odom_cb(self, msg):
        self.current_vel_x = msg.twist.twist.linear.x
    def module_status_cb(self, msg):
        self.module_status_msg = msg
        self.last_module_status = rospy.get_time()
        if not self.parse_json:
            return
        self.special_matching_error = ""
        self.special_matching_state = ""
        self.special_error_led = ""
        self.special_state_led = ""
        self.special_state_sound = ""
        self.special_error_sound = ""
        self.error_code_number = -1
        try:
            self.module_status_dict = json.loads(msg.data)
            # Normal error
            if "error_code" in self.module_status_dict:
                self.error_code = self.module_status_dict["error_code"]
            # Special error
            if python3:
                if "error_code" in self.module_status_dict and len(
                    self.error_list
                ):
                    for i in self.error_list:
                        if (
                            list(i.keys())[0]
                            in self.module_status_dict["error_code"]
                        ):
                            if "detail_status" in list(i.values())[0]:
                                self.special_matching_error = list(i.values())[0][
                                    "detail_status"
                                ]
                            if "reset_when_mission_error" in list(i.values())[0]:
                                self.reset_when_mission_error = list(i.values())[0][
                                    "reset_when_mission_error"
                                ]
                            else:
                                self.reset_when_mission_error = True
                            if "error_code_number" in list(i.values())[0]:
                                self.error_code_number = list(i.values())[0][
                                    "error_code_number"
                                ]
                            if "led_status" in list(i.values())[0]:
                                self.special_error_led = list(i.values())[0][
                                    "led_status"
                                ]
                            if "sound_status" in list(i.values())[0]:
                                self.special_error_sound = list(i.values())[0][
                                    "sound_status"
                                ]
                            return
                if "state" in self.module_status_dict and len(self.state_list):
                    for i in self.state_list:
                        if (
                            list(i.keys())[0]
                            in self.module_status_dict["state"]
                        ):
                            self.special_matching_state = list(i.values())[0][
                                "detail_status"
                            ]
                            if "led_status" in list(i.values())[0]:
                                self.special_state_led = list(i.values())[0][
                                    "led_status"
                                ]
                            if "sound_status" in list(i.values())[0]:
                                self.special_state_sound = list(i.values())[0][
                                    "sound_status"
                                ]
                            return
            else:
                if "error_code" in self.module_status_dict and len(
                    self.error_list
                ):
                    for i in self.error_list:
                        if i.keys()[0] in self.module_status_dict["error_code"]:
                            self.special_matching_error = i.values()[0][
                                "detail_status"
                            ]
                            if "led_status" in i.values()[0]:
                                self.special_error_led = i.values()[0][
                                    "led_status"
                                ]
                            if "sound_status" in i.values()[0]:
                                self.special_error_sound = i.values()[0][
                                    "sound_status"
                                ]
                            return
                if "state" in self.module_status_dict and len(self.state_list):
                    for i in self.state_list:
                        if i.keys()[0] in self.module_status_dict["state"]:
                            self.special_matching_state = i.values()[0][
                                "detail_status"
                            ]
                            if "led_status" in i.values()[0]:
                                self.special_state_led = i.values()[0][
                                    "led_status"
                                ]
                            if "sound_status" in i.values()[0]:
                                self.special_state_sound = i.values()[0][
                                    "sound_status"
                                ]
                            return
        except Exception as e:
            rospy.logerr("module_status_cb: {}".format(e))

    def _skip_check_type_vel(self):
        if "forward" in self.type_skip:
            if self.current_vel_x < self.skip_threshold:
                return True

        elif "backward" in self.type_skip:
            if self.current_vel_x > self.skip_threshold:
                return True
        return False


    def loop(self):
        r = rospy.Rate(2)
        self.vel_x = 0
        self.start_time_check_disconnect = rospy.get_time()
        while not rospy.is_shutdown():
            self.skip_check = False

            if "status_topic" in self.param_dict:
                if self.param_dict["status_topic"] == "/followline_sensor":
                    if self.vel_x == 0:
                        self.start_time_check_disconnect = rospy.get_time()
                        self.last_module_status = rospy.get_time()
                    if (rospy.get_time() - self.start_time_check_disconnect <= 0.5):
                        self.last_module_status = rospy.get_time()
            if not self.condition:
                self.skip_check = False
            elif self.condition and "vel" in self.type_skip:
                self.skip_check = self._skip_check_type_vel()
            if self.skip_check:
                self.last_module_status = rospy.get_time()
                self.module_alive = True
            else:
                if self.check_script == "":
                    if rospy.get_time() - self.last_module_status > self.timeout:
                        self.module_alive = False
                    else:
                        self.module_alive = True
                else:
                    # Ex: msg.ranges[0] == -1
                    if self.module_status_msg != None:
                        cmd = "self.module_alive = {}".format(self.check_script)
                        exec(cmd)
                    if rospy.get_time() - self.last_module_status > self.timeout:
                        self.module_alive = False
            r.sleep()
